### PR TITLE
[DO NOT MERGE] Adding custom errors for client credentials hook.

### PR DESCRIPTION
**Pull request by seejee**
Getting feedback on an approach to send user errors back from client credentials in an "envelope" that doesn't actually respond with a 500 status code.

### CODE REVIEW

File: lib/compilers/client_credentials_exchange.js [hunk 1]
```diff
@@ -3,12 +3,27 @@
 const Authz = require('../authorization');
 const Factory = require('./compilerFactory');
 
-module.exports = Factory.createCompiler(clientCredentialsExchangeHandler);
+class InvalidRequestError extends Error {
+    constructor(message) {
+        super(message);
+        this.name = "InvalidRequestError";
```
[the remark this answers is not in the document]
Comment (ggoodman) [reply lib/compilers/client_credentials_exchange.js:9]: Oh, but that might barf w/ the `prelude` approach 🤔

File: lib/compilers/client_credentials_exchange.js [hunk 2]
```diff
@@ -3,12 +3,27 @@
...
+    }
+}
+
+function addPrelude(script) {
+    return `
+        ${InvalidRequestError.toString()}
+        ${script}
+    `;
+}
```
Comment (ggoodman) [inline lib/compilers/client_credentials_exchange.js:18]: I think we could play around a bit here to get a more flexible approach but overall this is the right direction.

**Comment by seejee** [comment]
@ggoodman I incorporated some of your feedback. Would love to hear your thoughts before I roll out this pattern to the other hook compilers. Thanks!

File: lib/compilers/client_credentials_exchange.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (seejee) [reply lib/compilers/client_credentials_exchange.js]: I used this approach by adding BaseError to the prelude as well.

File: lib/compilers/client_credentials_exchange.js (hunk 2 above)
Comment (seejee) [reply lib/compilers/client_credentials_exchange.js]: Check out what I did in the latest. I'm passing in an array of functions to be added and letting the compilerFactory actually do the script munging.

File: lib/compilers/client_credentials_exchange.js [hunk 3]
```diff
@@ -3,12 +3,86 @@
...
+        data: {
+            error: 'server_error',
+            error_code: err.code,
+            error_description: err.message,
+            stack: err.stack,
+        }
+    })
+};
+
+const PRELUDE_ITEMS = [
```
Comment (ggoodman) [inline lib/compilers/client_credentials_exchange.js:62]: I think I understand what's going on in this approach. It seems to me that the 'prelude' injection means that the order of the items in this array is significant.

I'd like to see a comment explaining what the array represents and how it is intended to be used for when we come back a year from now.

File: lib/compilers/client_credentials_exchange.js [hunk 4]
```diff
@@ -3,12 +3,86 @@
...
+    BaseError,
+    InvalidRequestError,
+    InvalidScopeError,
+    ServerError
+];
```
Comment (ggoodman) [inline lib/compilers/client_credentials_exchange.js:67]:
```suggestion
// The array below is designed to be passed as the second argument to `createCompiler`. Each item
// in the array will be stringified (have `.toString()` called on it) and will be placed in a separate line from
// other prelude items. The actual script will appear after another newline following the prelude items.
const PRELUDE_ITEMS = [
    BaseError,
    InvalidRequestError,
    InvalidScopeError,
    ServerError
];
```

File: lib/compilers/compilerFactory.js [hunk 5]
```diff
@@ -13,15 +20,17 @@ module.exports = {
  * @param  {Function} handler The Extensibility Point handler
```
Comment (shawnmclean) [inline lib/compilers/compilerFactory.js:20]:
```suggestion
 * @param  {Function} handler The Extensibility Point handler
 * @param {Array} preludeItems Items that will be rendered (toString()) into the start of the script, separated by a new line.
```

File: package-lock.json [hunk 6]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "auth0-ext-compilers",
-  "version": "5.5.2",
+  "version": "5.5.3",
```
Comment (shawnmclean) [inline package-lock.json:3]: 👍 good catch on updating this. Now that we're here, should we bump the version in this PR?

File: lib/compilers/client_credentials_exchange.js [hunk 7]
```diff
@@ -3,12 +3,86 @@
...
+
+class ServerError extends BaseError {
+    constructor(code, message) {
+        super(code, message);
+    }
+}
+
+const ERROR_TO_RESULT_MAP =  {
+    [InvalidRequestError.name]: (err) => ({
+        result: 'oauth_error',
```
Comment (shawnmclean) [inline lib/compilers/client_credentials_exchange.js:34]: After reviewing https://github.com/auth0/auth0-ext/pull/111/files and saw how this is used, I think the name should match the name of the hook, that way, there would be some consistency/format for other hooks to follow. Something like below:
```suggestion
        result: 'client_credential_exchange_error',
```

**Comment by seejee** [comment]
closing in favor of #20

**Pull request closed without merging**